### DISCUSSION TITLE
microphone gain

### DISCUSSION THREAD

**Discussion by juha-h** [baresip/baresip#2668] [Q&A] [answered]
Some users of my Android app complain that microphone volume is too low. Android API does not have any means to control microphone gain.  So I started to wonder, could it be possible to increase microphone gain in baresip opensles module.

I found this text about PCM volume control:

https://web.archive.org/web/20150923161151/http://www.ypass.net/blog/2010/01/pcm-audio-part-3-basic-audio-effects-volume-control/

It tells:
```
So if you want to increase the volume of your PCM stream, just multiply every PCM value by some number.
If we had 2048 bytes of audio (remember… that’s 1024 samples since each sample is two bytes),
we could amplify the stream with this type of code:

int16_t pcm[1024] = read in some pcm data;
for (ctr = 0; ctr < 1024; ctr++) {
    pcm[ctr] *= 2;
}
```
Could that kind of multiplying be added to baresip opensles recorder.c and if so where?

**Comment by sreimers** [comment, marked as answer]
Using software gain should always be the last option (automatic or hardware gain control should be preferred). Since audio is clipping and distorted if gain is already to big. But you can simply add a custom baresip audio filter module (`aufilt_register`) to increase gain before encoding.

**Comment by juha-h** [comment]
Thanks for the suggestion.  I don't much about audio, but could try when I have time. Clipping issue as well as that volume as perceived by humans is logarithmic, was mentioned in the quoted article.

**Comment by assegaf** [comment]
yes looking for it also, 
but another options I tried from media server /sip server (asterisk), sound distorted, dont have something like normalization sound volume level, so I am looking microphone gain in android and has something like normalization at exact value sound not distorted.

**Comment by juha-h** [comment]
@assegaf Did you implement an audio filter module for microphone gain?
